File: py/mds/introspect.py
"""Introspection and generation over schema contracts (sections 47/48):
machine-readable inspection reports and Markdown skeleton generation.

Pure library functions - no argv handling, no console I/O. The CLI
(:mod:`mds.cli`) is a thin wrapper around these plus :mod:`mds.validate`.

Mirrors ``js/src/introspect.js``.
"""

from .validate import load_schema


def _render_expect(out, indent, text):
    if not text:
        return
    out.append(f"{indent}- expect:")
    for line in str(text).split("\n"):
        out.append(f"{indent}      {line}")


def _render_validate(out, indent, holder):
    """Render a region's validate binding, if any."""
    vs = holder.get("validateSem")
    if not vs:
        return
    out.append(f"{indent}- validate semantic: {vs['semantic']}")


def inspect_schema(schema_text, schema_name="schema.mds", base_dir="."):
    """Render an introspection report (Markdown list) for a schema.

    Returns ``{"exitCode": int, "stream": str}``.
    """
    loaded = load_schema(schema_text, schema_name, base_dir)
    model, diags = loaded["model"], loaded["diags"]
    if any(d.severity == "error" for d in diags):
        return {"exitCode": 2, "stream": "\n".join(d.render() for d in diags)}
    out = []
    out.append(f"- document: {model['documentName'] or '(unnamed)'}")
    out.append(f"- order: {model['orderMode']}")
    out.append(f"- additionalSections: {str(model['additionalSections']).lower()}")
    out.append(f"- additionalFields: {str(model['additionalFields']).lower()}")
    _render_expect(out, "", model.get("expect"))
    _render_validate(out, "", model)

    def walk(heads, indent):
        for h in heads:
            card = "" if h["card"] == "required" else f" [{h['card']}]"
            out.append(f"{indent}- ## {h['label']}{card}")
            _render_expect(out, f"{indent}  ", h.get("expect"))
            _render_validate(out, f"{indent}  ", h)
            if h.get("prose"):
                out.append(f"{indent}  - prose {h['prose']['card']}")
                _render_expect(out, f"{indent}    ", h["prose"].get("expect"))
                _render_validate(out, f"{indent}    ", h["prose"])
            if h.get("list"):
                tail = f"{indent}  - list {h['list']['typeExpr']} {h['list']['card']}".rstrip()
                out.append(tail)
                _render_expect(out, f"{indent}    ", h["list"].get("expect"))
                _render_validate(out, f"{indent}    ", h["list"])
            for t in h["tables"]:
                cols = ", ".join(c["label"] for c in t["cols"])
                out.append(f"{indent}  - table {t.get('name') or '(anon)'} ({cols})")
                _render_expect(out, f"{indent}    ", t.get("expect"))
                for c in t["cols"]:
                    if c.get("expect"):
                        out.append(f"{indent}    - column {c['label']}:")
                        _render_expect(out, f"{indent}      ", c.get("expect"))
            for e in h["embeds"]:
                out.append(f"{indent}  - embed {e['format']} {e['card']}")
                _render_expect(out, f"{indent}    ", e.get("expect"))
            for f in h["fields"]:
                tail = (f"{indent}  - field {f.get('id') or f['label']}: "
                        f"{f['typeExpr']} {f['card']}").rstrip()
                out.append(tail)
                _render_expect(out, f"{indent}    ", f.get("expect"))
            walk(h["children"], f"{indent}    ")

    walk(model["sections"], "")
    return {"exitCode": 0, "stream": "\n".join(out)}


def scaffold_doc(schema_text, schema_name="schema.mds", base_dir="."):
    """Generate a Markdown skeleton satisfying the section structure.

    Returns ``{"exitCode": int, "stream": str}``.
    """
    loaded = load_schema(schema_text, schema_name, base_dir)
    model, diags = loaded["model"], loaded["diags"]
    if any(d.severity == "error" for d in diags):
        return {"exitCode": 2, "stream": "\n".join(d.render() for d in diags)}
    out = []

    def walk(heads):
        for h in heads:
            out.append("")
            label = "<Title>" if h["glob"] else h["label"]
            out.append(f"{'#' * h['level']} {label}")
            out.append("")
            if h.get("prose"):
                out.append("...")
            if h.get("list"):
                out.append("- ...")
            for t in h["tables"]:
                out.append("")
                out.append("| " + " | ".join(c["label"] for c in t["cols"]) + " |")
                out.append("|" + "|".join("---" for _ in t["cols"]) + "|")
            for e in h["embeds"]:
                out.append("")
                out.append(f"```{e['format']}")
                out.append("...")
                out.append("```")
            for f in h["fields"]:
                if not f.get("children") and f["card"] != "zero-or-more":
                    out.append(f"- {f['label']}: ")
            walk(h["children"])

    walk(model["sections"])
    return {"exitCode": 0, "stream": "\n".join(out).lstrip("\n")}

File: js/src/introspect.js
/**
 * Introspection and generation over schema contracts (sections 47/48):
 * machine-readable inspection reports and Markdown skeleton generation.
 *
 * Pure library functions — no argv handling, no process I/O. The CLI
 * (`cli.js`) is a thin wrapper around these plus {@link module:validate}.
 *
 * @module introspect
 */
import { loadSchema } from './validate.js';

/** Render an indented `expect:` block into an introspection report. */
function pushExpect(out, indent, text) {
  if (!text) return;
  out.push(`${indent}- expect:`);
  for (const l of String(text).split('\n')) out.push(`${indent}      ${l}`);
}

/** Render a region's validate binding, if any. */
function pushValidate(out, indent, holder) {
  if (!holder.validateSem) return;
  out.push(`${indent}- validate semantic: ${holder.validateSem.semantic}`);
}

/**
 * Render an introspection report (Markdown list) for a schema.
 *
 * @param {string} schemaText raw `.mds` source
 * @param {string} [schemaName='schema.mds'] file name used in diagnostics
 * @param {string} [baseDir=process.cwd()] resolves schema imports
 * @returns {{exitCode:number, stream:string}}
 */
export function inspectSchema(schemaText, schemaName = 'schema.mds', baseDir = process.cwd()) {
  const { model, diags } = loadSchema(schemaText, schemaName, baseDir);
  if (diags.some((d) => d.severity === 'error')) {
    return { exitCode: 2, stream: diags.map((d) => d.render()).join('\n') };
  }
  const out = [];
  out.push(`- document: ${model.documentName ?? '(unnamed)'}`);
  out.push(`- order: ${model.orderMode}`);
  out.push(`- additionalSections: ${model.additionalSections}`);
  out.push(`- additionalFields: ${model.additionalFields}`);
  pushExpect(out, '', model.expect);
  pushValidate(out, '', model);
  const walk = (heads, indent) => {
    for (const h of heads) {
      const card = h.card === 'required' ? '' : ` [${h.card}]`;
      out.push(`${indent}- ## ${h.label}${card}`);
      pushExpect(out, `${indent}  `, h.expect);
      pushValidate(out, `${indent}  `, h);
      if (h.prose) {
        out.push(`${indent}  - prose ${h.prose.card}`);
        pushExpect(out, `${indent}    `, h.prose.expect);
        pushValidate(out, `${indent}    `, h.prose);
      }
      if (h.list) {
        out.push(`${indent}  - list ${h.list.typeExpr} ${h.list.card}`.trimEnd());
        pushExpect(out, `${indent}    `, h.list.expect);
        pushValidate(out, `${indent}    `, h.list);
      }
      for (const t of h.tables) {
        out.push(`${indent}  - table ${t.name ?? '(anon)'} (${t.cols.map((c) => c.label).join(', ')})`);
        pushExpect(out, `${indent}    `, t.expect);
        for (const c of t.cols) {
          if (c.expect) {
            out.push(`${indent}    - column ${c.label}:`);
            pushExpect(out, `${indent}      `, c.expect);
          }
        }
      }
      for (const e of h.embeds) {
        out.push(`${indent}  - embed ${e.format} ${e.card}`);
        pushExpect(out, `${indent}    `, e.expect);
      }
      for (const f of h.fields) {
        out.push(`${indent}  - field ${f.id ?? f.label}: ${f.typeExpr} ${f.card}`.trimEnd());
        pushExpect(out, `${indent}    `, f.expect);
      }
      walk(h.children, `${indent}    `);
    }
  };
  walk(model.sections, '');
  return { exitCode: 0, stream: out.join('\n') };
}

/**
 * Generate a Markdown skeleton satisfying the section structure.
 *
 * @param {string} schemaText raw `.mds` source
 * @param {string} [schemaName='schema.mds'] file name used in diagnostics
 * @param {string} [baseDir=process.cwd()] resolves schema imports
 * @returns {{exitCode:number, stream:string}}
 */
export function scaffoldDoc(schemaText, schemaName = 'schema.mds', baseDir = process.cwd()) {
  const { model, diags } = loadSchema(schemaText, schemaName, baseDir);
  if (diags.some((d) => d.severity === 'error')) {
    return { exitCode: 2, stream: diags.map((d) => d.render()).join('\n') };
  }
  const out = [];
  const hashes = (n) => '#'.repeat(n);
  const walk = (heads) => {
    for (const h of heads) {
      out.push('');
      const label = h.glob ? '<Title>' : h.label;
      out.push(`${hashes(h.level)} ${label}`);
      out.push('');
      if (h.prose) out.push('...');
      if (h.list) out.push('- ...');
      for (const t of h.tables) {
        out.push('');
        out.push(`| ${t.cols.map((c) => c.label).join(' | ')} |`);
        out.push(`|${t.cols.map(() => '---').join('|')}|`);
      }
      for (const e of h.embeds) {
        out.push('');
        out.push(`\`\`\`${e.format}`);
        out.push('...');
        out.push('```');
      }
      for (const f of h.fields) {
        if ((f.children?.length ?? 0) === 0 && f.card !== 'zero-or-more') {
          out.push(`- ${f.label}: `);
        }
      }
      walk(h.children);
    }
  };
  walk(model.sections);
  return { exitCode: 0, stream: out.join('\n').replace(/^\n+/, '') };
}
